### PR TITLE
WIP Move express to fastify

### PR DESCRIPTION
**Pull request by jackdclark**
- [x] Move all routes to fastify
- [x] Move middleware to fastify
- [x] Add request route validation / schema
- [ ] Add response route validation / schema
- [ ] Move to [fastify-mongodb](https://github.com/fastify/fastify-mongodb)
- [ ] Run benchmarks

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -62,64 +62,55 @@ if(process.env.VCAP_SERVICES){
 }
 logger.info("db type=="+dbtype);
 
-var routes = new require('./routes/index.js')(dbtype, authService,settings);
+var routes = new require('./routes/index.js')(dbtype, authService, settings);
```
Comment (mcollina) [inline app.js:65]: I would not define the routes this way in a proper porting. This is not a good example on how to structure _any_ Node.js applications (unfortunately). Check out https://github.com/fastify/fastify-example-twitter as structure example.
You can log this as a separate issue, as it's a refactoring that does not impact performance at all, but rather it's a more proper application structure.

File: app.js [hunk 2]
```diff
@@ -62,64 +62,55 @@ if(process.env.VCAP_SERVICES){
...
+	fastify.get('/config/countFlights', {}, routes.countFlights);
+	fastify.get('/config/countFlightSegments', {}, routes.countFlightSegments);
+	fastify.get('/config/countAirports', {}, routes.countAirports);
+	// fastify.get('/loaddb', startLoadDatabase);
+	fastify.get('/loader/load', {}, startLoadDatabase);
+	fastify.get('/loader/query', {}, loader.getNumConfiguredCustomers);
+	fastify.get('/checkstatus', {}, checkStatus);
+
+	next()
+}
```
Comment (mcollina) [inline app.js:103]: The indentation of your editor uses both spaces and tabs. You might want to run everything through eslint or standard (or semistandard).

File: app.js [hunk 3]
```diff
@@ -134,61 +125,66 @@ if (authService && authService.initialize)
...
+function checkStatus(req, reply) {
+	reply.send("OK");
 }
 
-function login(req, res){
-	if (!initialized)
-     {
+function login(req, reply) {
+	if (!initialized) {
 		logger.info("please wait for db connection initialized then trigger again.");
```
Comment (mcollina) [inline app.js:134]: you can access fastify logger instead.

File: app.js [hunk 4]
```diff
@@ -134,61 +125,66 @@ if (authService && authService.initialize)
...
-		logger.info('Error connecting to database - exiting process: '+ error);
-		// Do not stop the process for debug in container service
-		//process.exit(1); 
-	}else
-	      initialized =true;
-
-	logger.info("Initialized database connections");
-	startServer();
+	if (initialized ) return;
+	routes.initializeDatabaseConnections(function(error) {
```
Comment (mcollina) [inline app.js:165]: this is not a good pattern. The routes should not have this type of init phase, but rather you should have a plugin that provides the database connection to them. You can also use fastify-mongodb.

File: app.js [hunk 5]
```diff
@@ -134,61 +125,66 @@ if (authService && authService.initialize)
...
 	serverStarted = true;
-	app.listen(port);   
-	logger.info("Express server listening on port " + port);
+	fastify.listen(port, function (err) {
+    if (err) {
+      logger.error("Error starting server " + err);
+      process.exit(1)
+    }
+  });
+	logger.info("Fastify server listening on port " + port);
```
Comment (mcollina) [inline app.js:189]: The whole init sequence is not behaving correctly.

What you want is:

```js
if (process.main === module) {
  start() // this build the instance and starts the app
}
```

**Review by mcollina — COMMENTED** [review]
Good work. I've left some notes to improve things around here. You can completely skip them if you want, those are more suggestions to improve the code quality of acmeair.

**Comment by jackdclark** [comment]
Closing this as #2 includes all of these commits

**Pull request closed without merging**